README.rst
.. image:: https://img.shields.io/badge/License-AGPL%20v3-blue.svg
   :target: https://www.gnu.org/licenses/agpl-3.0

=======
Key-CNN
=======

Key-CNN is a simple CNN-based framework for estimating harmonic properties
of music tracks.

First and foremost, Key-CNN is a key estimator. To determine the global key of
an audio file, simply run the script

.. code-block:: console

    key -i my_audio.wav

You may specify other models and output formats (`mirex <https://www.music-ir.org/mirex/wiki/2019:Audio_Key_Detection>`_,
`JAMS <https://github.com/marl/jams>`_) via command line parameters.

E.g. to create JAMS as output format and a deepsquare model used in the SMC
paper [1], please run

.. code-block:: console

    key -m deepsquare --jams -i my_audio.wav


To use one of the ``DeepSpec`` models from [1] (see also repo
`directional_cnns <https://github.com/hendriks73/directional_cnns>`_), run

.. code-block:: console

    key -m deepspec --jams -i my_audio.wav

or,

.. code-block:: console

    key -m deepspec_k24 --jams -i my_audio.wav

if you want to use a higher capacity model (some ``k``-values are supported).
``deepsquare`` and ``shallowspec`` models may also be used.

For estimation using models trained for [2], you may run one of the following
model specs:

.. code-block:: console

    key -m winterreise -i my_audio.wav
    key -m winterreise_v -i my_audio.wav
    key -m winterreise_s -i my_audio.wav
    key -m winterreise_v_fold0 -i my_audio.wav
    key -m winterreise_s_fold1 -i my_audio.wav

For more model names and split training split definitions, please see the `models directory
in the GitHub repo <https://github.com/hendriks73/key-cnn/tree/master/keycnn/models>`_
(just remove the ``.h5`` from the file name to use as model name).
The groundtruth annotations for Winterreise models may be found
`here <https://github.com/hendriks73/key-cnn/tree/master/annotations/winterreise>`_.

For batch processing, you may want to run ``key`` like this:

.. code-block:: console

    find /your_audio_dir/ -name '*.wav' -print0 | xargs -0 key -d /output_dir/ -i

This will recursively search for all ``.wav`` files in ``/your_audio_dir/``, analyze then
and write the results to individual files in ``/output_dir/``. Because the model is only
loaded once, this method of processing is much faster than individual program starts.

Instead of estimating a global key, Key-CNN can also estimate local keys in the
form of a keygram. This can be useful for identifying modulations.
To create such a keygram, run

.. code-block:: console

    keygram -p my_audio.wav

As output, ``keygram`` will create a ``.png`` file. Additional options to select different models
and output formats are available.

You may use the ``--csv`` option to export local key estimates in a parseable format and the
``--hop-length`` option to change temporal resolution.
The parameters ``--sharpen`` and ``--norm-frame`` let you post-process the image.


Installation
============

Clone this repo and run ``setup.py install`` using Python 3.6:

.. code-block:: console

    git clone https://github.com/hendriks73/key-cnn.git
    cd key-cnn
    python setup.py install

You may need to install TensorFlow using ``pip`` from the command line.

License
=======

Source code and models can be licensed under the GNU AFFERO GENERAL PUBLIC LICENSE v3.
For details, please see the `LICENSE <LICENSE>`_ file.


Citation
========

If you use Key-CNN in your work, please consider citing it.
ShallowSpec, DeepSpec, and DeepSquare models:

.. code-block:: latex

   @inproceedings{SchreiberM19_CNNKeyTempo_SMC,
      Title = {Musical Tempo and Key Estimation using Convolutional Neural Networks with Directional Filters},
      Author = {Hendrik Schreiber and Meinard M{\"u}ller},
      Booktitle = {Proceedings of the Sound and Music Computing Conference ({SMC})},
      Pages = {47--54},
      Year = {2019},
      Address = {M{\'a}laga, Spain}
   }


All Winterreise [2] models and annotations:

.. code-block:: latex

   @inproceedings{SchreiberWM20_HMMCNNLocalKey_ICASSP,
      Title = {Local Key Estimation in Classical Music Recordings: A Cross-Version Study on {Schubert's} {Winterreise}},
      Author = {Hendrik Schreiber, Christof Wei{\ss}, Meinard M{\"u}ller},
      Booktitle = {Proceedings of the {IEEE} International Conference on Acoustics, Speech, and Signal Processing ({ICASSP})},
      Year = {2020},
      Address = {Barcelona, Spain}
   }

References
==========

.. [1] Hendrik Schreiber, Meinard Müller, `Musical Tempo and Key Estimation using Convolutional
    Neural Networks with Directional Filters
    <http://smc2019.uma.es/articles/P1/P1_07_SMC2019_paper.pdf>`_
    Proceedings of the Sound and Music Computing Conference (SMC),
    Málaga, Spain, 2019.
.. [2] Hendrik Schreiber, Christof Weiß, Meinard Müller,
    "Local Key Estimation in Classical Music Recordings: A Cross-Version Study on Schubert's Winterreise."
    Proceedings of the IEEE International Conference on Acoustics, Speech, and Signal Processing (ICASSP),
    Barcelona, Spain, 2020.


CHANGES.rst
=======
Changes
=======

0.0.2:
 - Added Winterreise local key models and annotations.
 - Updated TensorFlow dependency to v1.15.2.

0.0.1:
 - Initial version

tests/test_feature.py
import unittest

import librosa

from keycnn.feature import read_features


class TestTempoClassifier(unittest.TestCase):

    def test_init(self):
        file = 'data/drumtrack.mp3'
        y, sr = librosa.load(file, sr=22050)
        # possible features frames
        num_feature_frames = y.shape[0] / 4096
        # possible feature windows with half overlap
        num_feature_windows = (num_feature_frames // 30) // 2

        features = read_features(file)
        self.assertEqual(len(features.shape), 4)
        self.assertEqual(features.shape[0], num_feature_windows)
        self.assertEqual(features.shape[1], 168)
        self.assertEqual(features.shape[2], 60)
        self.assertEqual(features.shape[3], 1)



tests/test_classifier.py
import unittest

import numpy as np

from keycnn.classifier import KeyClassifier
from keycnn.feature import read_features


class TestKeyClassifier(unittest.TestCase):

    def test_init(self):
        key_classifier = KeyClassifier('deepspec')
        self.assertIsNotNone(key_classifier.model)

    def test_winterreise_init(self):
        key_classifier = KeyClassifier('winterreise')
        self.assertIsNotNone(key_classifier.model)

    def test_bad_model_name(self):
        try:
            KeyClassifier('bad_model_name')
            self.fail('Expected FileNotFoundError')
        except FileNotFoundError:
            pass

    def test_predict(self):
        key_classifier = KeyClassifier('deepspec')
        predictions = key_classifier.estimate(self.get_test_data())
        self.assertEqual(predictions.shape, (2, 24))
        np.testing.assert_array_almost_equal(np.ones(2), np.sum(predictions, axis=1))
        tempi = np.argmax(predictions, axis=1)
        self.assertEqual(tempi[0], 5)
        self.assertEqual(tempi[1], 2)

    def test_predict_key(self):
        key_classifier = KeyClassifier('deepspec')
        key = key_classifier.estimate_key(self.get_test_data())
        self.assertEqual('D', key[0])
        self.assertEqual('major', key[1])

    def test_key_with_real_data(self):
        features = read_features('data/drumtrack.mp3')
        key_classifier = KeyClassifier('deepspec')
        key = key_classifier.estimate_key(features)
        self.assertEqual('C', key[0])
        self.assertEqual('major', key[1])

    def get_test_data(self):
        artificial_data = np.zeros((2, 168, 60, 1))
        for i in np.arange(0, 60, 4):
            artificial_data[0, :, i, 0] = 1
        for i in np.arange(0, 60, 30):
            artificial_data[1, :, i, 0] = 1
        return artificial_data


keycnn/feature.py
# encoding: utf-8

"""
Feature loading from audio files.

Specifically, key-cnn uses CQT spectra with 8 octaves starting at C1.
"""

import librosa as librosa
import numpy as np


def read_features(file, frames=60, hop_length=30, zero_pad=False):
    """
    Resample file to 22050 Hz, then transform using CQT with length 8192
    and hop size 4096, ranging from E1 + 7 octaves with two semitones
    per bin.

    Since we require at least 60 frames, shorter audio excerpts are always
    zero padded.

    Specifically for keygram, 30 frames each can be added at the front and
    at the back in order to make the calculation of key values for the first
    and the last window possible.

    :param file: file
    :param frames: 60
    :param hop_length: 30 or shorter
    :param zero_pad: adds 30 zero frames both at the front and back
    :return: feature tensor for the whole file
    """
    y, sr = librosa.load(file, sr=22050)
    octaves = 7
    bins_per_semitone = 2
    bins_per_octave = 12 * bins_per_semitone
    window_length = 8192
    data = np.abs(librosa.cqt(y, sr=sr, hop_length=window_length // 2,
                              fmin=librosa.note_to_hz('E1'),
                              n_bins=bins_per_octave * octaves,
                              bins_per_octave=bins_per_octave))

    data = np.reshape(data, (1, data.shape[0], data.shape[1], 1))

    # add frames/2 zero frames before and after the data
    if zero_pad:
        data = _add_zeros(data, frames)

    # zero-pad, if we have less than 256 frames to make sure we get some
    # result at all
    if data.shape[2] < frames:
        data = _ensure_length(data, frames)

    # convert data to overlapping windows,
    # each window is one sample (first dim)
    return _to_sliding_window(data, frames, hop_length)


def _ensure_length(data, length):
    padded_data = np.zeros((1, data.shape[1], length, 1), dtype=data.dtype)
    padded_data[0, :, 0:data.shape[2], 0] = data[0, :, :, 0]
    return padded_data


def _add_zeros(data, zeros):
    padded_data = np.zeros((1, data.shape[1], data.shape[2] + zeros, 1), dtype=data.dtype)
    padded_data[0, :, zeros // 2:data.shape[2] + (zeros // 2), 0] = data[0, :, :, 0]
    return padded_data


def _to_sliding_window(data, window_length, hop_length):
    total_frames = data.shape[2]
    windowed_data = []
    for offset in range(0, ((total_frames - window_length) // hop_length + 1) * hop_length, hop_length):
        windowed_data.append(np.copy(data[:, :, offset:window_length + offset, :]))
    return np.concatenate(windowed_data, axis=0)


keycnn/classifier.py
# encoding: utf-8

import os
import pkgutil
import sys
import tempfile

import librosa
import numpy as np
from tensorflow.python.keras.models import load_model


def std_normalizer(data):
    """
    Normalizes data to zero mean and unit variance.

    :param data: data
    :return: standardized data
    """
    # normalize as 64 bit, to avoid numpy warnings
    data = data.astype(np.float64)
    mean = np.mean(data)
    std = np.std(data)
    if std != 0.:
        data = (data-mean) / std
    return data.astype(np.float16)


class KeyClassifier:
    """
    Classifier that can estimate musical key in different formats.
    """

    def __init__(self, model_name='deepspec'):
        """
        Initializes this classifier with a Keras model.

        :param model_name: model name from sub-package models. E.g. 'deepspec', 'shallowspec', or 'deepsquare'
        """

        def to_major_minor_key(index):

            if not np.isscalar(index):
                return [to_major_minor_key(x) for x in index]

            minor = index >= 12
            midi = index + 12
            if minor:
                midi = index - 12
            tonic = librosa.midi_to_note(midi=midi, octave=False)
            mode = 'minor' if minor else 'major'
            return tonic, mode

        self.to_key = to_major_minor_key

        # aliases for specific deep/shallow models
        aliases = {
            'deepspec': 'deepspec_k16',
            'shallowspec': 'shallowspec_k4',
            'deepsquare': 'deepsquare_k8',
            'winterreise': 'ds_winterreise_v_fold0_HU33_OL06_QU98_SC06_TR99',
            'winterreise_v': 'ds_winterreise_v_fold0_HU33_OL06_QU98_SC06_TR99',
            'winterreise_v_fold0': 'ds_winterreise_v_fold0_HU33_OL06_QU98_SC06_TR99',
            'winterreise_v_fold1': 'ds_winterreise_v_fold1_SC06_TR99_AL98_FI55_FI66',
            'winterreise_v_fold2': 'ds_winterreise_v_fold2_FI55_FI66_FI80_HU33_OL06',
            'winterreise_s': 'ds_winterreise_s_fold0_12_13_14_15_16_17_18_19_20_21_22_23_24',
            'winterreise_s_fold0': 'ds_winterreise_s_fold0_12_13_14_15_16_17_18_19_20_21_22_23_24',
            'winterreise_s_fold1': 'ds_winterreise_s_fold1_20_21_22_23_24_01_02_03_04_05_06_07_08',
            'winterreise_s_fold2': 'ds_winterreise_s_fold2_04_05_06_07_08_09_10_11_12_13_14_15_16',
            'winterreise_n': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_01_02_03_FI80_HU33_04_05',
            'winterreise_n_fold00': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_01_02_03_FI80_HU33_04_05',
            'winterreise_n_fold01': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_04_05_06_FI80_HU33_07_08',
            'winterreise_n_fold02': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_07_08_09_FI80_HU33_10_11',
            'winterreise_n_fold03': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_10_11_12_FI80_HU33_13_14',
            'winterreise_n_fold04': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_13_14_15_FI80_HU33_16_17',
            'winterreise_n_fold05': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_16_17_18_FI80_HU33_19_20',
            'winterreise_n_fold06': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_19_20_21_FI80_HU33_22_23',
            'winterreise_n_fold07': 'ds_winterreise_n_fold_NOT_AL98_FI55_FI66_22_23_24_FI80_HU33_01_02',
            'winterreise_n_fold08': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_01_02_03_QU98_SC06_04_05',
            'winterreise_n_fold09': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_04_05_06_QU98_SC06_07_08',
            'winterreise_n_fold10': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_07_08_09_QU98_SC06_10_11',
            'winterreise_n_fold11': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_10_11_12_QU98_SC06_13_14',
            'winterreise_n_fold12': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_13_14_15_QU98_SC06_16_17',
            'winterreise_n_fold13': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_16_17_18_QU98_SC06_19_20',
            'winterreise_n_fold14': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_19_20_21_QU98_SC06_22_23',
            'winterreise_n_fold15': 'ds_winterreise_n_fold_NOT_FI80_HU33_OL06_22_23_24_QU98_SC06_01_02',
            'winterreise_n_fold16': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_01_02_03_AL98_FI55_04_05',
            'winterreise_n_fold17': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_04_05_06_AL98_FI55_07_08',
            'winterreise_n_fold18': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_07_08_09_AL98_FI55_10_11',
            'winterreise_n_fold19': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_10_11_12_AL98_FI55_13_14',
            'winterreise_n_fold20': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_13_14_15_AL98_FI55_16_17',
            'winterreise_n_fold21': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_16_17_18_AL98_FI55_19_20',
            'winterreise_n_fold22': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_19_20_21_AL98_FI55_22_23',
            'winterreise_n_fold23': 'ds_winterreise_n_fold_NOT_QU98_SC06_TR99_22_23_24_AL98_FI55_01_02',
        }
        if model_name in aliases:
            model_name = aliases[model_name]
        self.model_name = model_name

        self.normalize = std_normalizer

        resource = _to_model_resource(model_name)
        try:
            file = _extract_from_package(resource)
        except Exception as e:
            print('Failed to find a model named \'{}\'. Please check the model name.'.format(model_name),
                  file=sys.stderr)
            raise e
        try:
            self.model = load_model(file)
        finally:
            os.remove(file)

    def estimate(self, data):
        """
        Estimate a key distribution.
        Probabilities are indexed, starting with 30 BPM and ending with 286 BPM.

        :param data: features
        :return: key probability distribution
        """
        assert len(data.shape) == 4, 'Input data must be four dimensional. Actual shape was ' + str(data.shape)
        assert data.shape[1] == 168, 'Second dim of data must be 168. Actual shape was ' + str(data.shape)
        assert data.shape[2] == 60, 'Third dim of data must be 60. Actual shape was ' + str(data.shape)
        assert data.shape[3] == 1, 'Fourth dim of data must be 1. Actual shape was ' + str(data.shape)
        norm_data = self.normalize(data)
        return self.model.predict(norm_data, norm_data.shape[0])

    def estimate_key(self, data):
        """
        Estimates the pre-dominant global key.

        :param data: features
        :return: a single key
        """
        prediction = self.estimate(data)
        averaged_prediction = np.average(prediction, axis=0)
        index = np.argmax(averaged_prediction)
        return self.to_key(index)


def _to_model_resource(model_name):
    file = model_name
    if not model_name.endswith('.h5'):
        file = file + '.h5'
    if not file.startswith('models/'):
        file = 'models/' + file
    return file


def _extract_from_package(resource):
    data = pkgutil.get_data('keycnn', resource)
    with tempfile.NamedTemporaryFile(prefix='model', suffix='.h5', delete=False) as f:
        f.write(data)
        name = f.name
    return name
